### ISSUE TITLE
Sandstorm - generation of unique usernames

### ISSUE THREAD

**Issue by mitar** [wekan/wekan#574]
I was looking at the `sandstorm.js` code and I noticed that the approach you are using to generate unique username is not the best. You should probably use [`Accounts.findUserByUsername`](http://docs.meteor.com/#/full/Accounts-findUserByUsername) because usernames are now in Meteor case insensitive (but preserving). If you just search normally, you will be doing a case sensitive search.

Moreover, you probably want to just try to update and see if non-unique value exception is thrown, or at least allow this to happen and retry with next value. Currently you have a slight race condition there.

**Comment by mquandalle** [comment]
I’m unsure about using `Accounts.findUserByUsername` simply because it is shipped by the `accounts-password` package which we shouldn’t depend on Sandstorm, and because anyway this method isn’t that useful as an abstraction and we could write the query ourselves like we do everywhere else in the app. Am I right thinking that the following change would fix the issue you are raising?

```diff
--- a/sandstorm.js
+++ b/sandstorm.js
@@ -112,7 +112,7 @@ if (isSandstorm && Meteor.isServer) {
     let appendNumber = 0;
     while (Users.findOne({
       _id: { $ne: doc._id },
-      username: generateUniqueUsername(username, appendNumber),
+      username: RegExp(`^${generateUniqueUsername(username, appendNumber)}$`, 'i'),
     })) {
       appendNumber += 1;
     }
```

The RegExp would be pretty inefficient but that’s should be OK considering that the long-term solution is that Sandstorm will provide a way to remove this hack as explained in https://github.com/sandstorm-io/sandstorm/issues/1027#issuecomment-148509709.

**Comment by mitar** [comment]
Hm, you might have to escape that username. Potentially they can include some regular expression characters? Or do you cleanup username/handle values somewhere? Probably we can trust Sandstorm to pass valid handles in?

I think this is good.

There is still a race-condition though. You should probably wrap this into try-catch and check for E11000 duplicate key exception and go to next number if it is thrown.

**Comment by mquandalle** [comment]
You are right the username isn’t sanitized so this isn’t mergable as is. Could you expend on the race condition, I’m not sure I get it?

**Comment by mitar** [comment]
> You are right the username isn’t sanitized so this isn’t mergable as is.

It seems it is OK, from Sandstorm documentation: It contains only lower-case ASCII letters, digits, and underscores, and it never starts with a digit. 

So we can hope that Sandstorm sanitizes this and just go with it.

> Could you expend on the race condition, I’m not sure I get it?

You are doing first queries, and then when you find an username which is not in the database, you make another query which sets the username. Now, between you finding an username and you setting the username there might be some other session somewhere else which just did the same for same username. So your username setting will fail because of unique index on one of those.

This is low probability (like all race conditions are), but it can happen.

There are two steps here:
* you wrap it inside the try/catch and you catch an exception of duplicate key, and you just retry another number
* because usernames are case-insensitive, there might be that two usernames which differ just in case were added at the same time and key would not match them; what you should then do is after update check if there is really just one username with case-insensitive regex in the database, if there is more than two, you revert your change and go for the next number

It is a bit complicated, this is handled by those Meteor username methods.

The question is if you really want to do it. I just pointed out that there is a race condition. Because they are rare you might not invest time into fixing it. But once you know how to see and fix them it is pretty straightforward to do it here and in the future. Is pretty common pattern.

**Comment by xet7** [comment]
@ocdtrekkie

Does this need fixing on Sandstorm?

**Comment by ocdtrekkie** [comment]
Do not know/am not familiar that with the sandstorm.js contents of Wekan.
